ipatests/test_integration/test_hbac_functional.py

```diff
@@ -0,0 +1,4558 @@
+            ["ipa", "hbacrule-add-service", self.HBAC_RULE,
+             "--hbacsvcs=sshd"]
+        )
+        self.master.run_command(
+            ["ipa", "hbacrule-show", self.HBAC_RULE, "--all"]
+        )
+
+        # Test access
+        # Test with hbactest - Basic tests
+        self.run_hbactest(self.USER_1, self.clients[1].hostname, "sshd",
+                          should_allow=True)
+        self.run_hbactest(self.USER_2, self.clients[1].hostname, "sshd",
+                          should_allow=False)
+        self.run_hbactest(
+            self.USER_1, "externalhost.randomhost.com", "sshd",
+            should_allow=False
+        )
+        self.run_hbactest(
+            self.USER_1, self.clients[1].hostname, "sshd",
+            self.HBAC_RULE,
+            expected_patterns=[
+                r"Access granted: True",
+                rf"Matched rules: {self.HBAC_RULE}"
+            ]
+        )
+        self.run_hbactest(
+            self.USER_1, self.master.hostname, "sshd",
+            self.HBAC_RULE,
+            expected_patterns=[
+                r"Access granted: False",
+                rf"Not matched rules: {self.HBAC_RULE}"
+            ]
+        )
+        self.run_hbactest(
+            self.USER_2, "externalhost.randomhost.com", "sshd",
+            self.HBAC_RULE,
+            expected_patterns=[
+                r"Access granted: False",
+                rf"Not matched rules: {self.HBAC_RULE}"
+            ]
+        )
+        self.run_hbactest(
+            self.USER_1, self.clients[1].hostname, "sshd",
+            should_allow=True, rule=self.HBAC_RULE, nodetail=True
+        )
+        self.run_hbactest(
+            self.USER_1, self.clients[1].hostname, "sshd",
+            rule=self.HBAC_RULE, nodetail=True,
+            forbidden_patterns=[rf"Matched rules: {self.HBAC_RULE}"]
+        )
+
+    # Test 025: Group access to client from external host 2
+    def test_hbacsvc_master_025(self, request):
+        """hbacsvc_master_025: group access client from external host 2,
+        on master, add rule and run hbactest"""
+        request.addfinalizer(lambda: self.cleanup_resources(
+            hbacrules=[self.HBAC_RULE],
+            groups=[self.GROUP, self.GROUP_2]
+        ))
+        tasks.kinit_admin(self.master)
+        self.master.run_command(
+            ["ipa", "group-add", self.GROUP, f"--desc={self.GROUP}"]
+        )
+        self.master.run_command(
+            ["ipa", "group-add", self.GROUP_2, f"--desc={self.GROUP_2}"]
+        )
+        self.master.run_command(
+            ["ipa", "group-add-member", self.GROUP,
+             f"--users={self.USER_1}"]
+        )
+        self.master.run_command(
+            ["ipa", "group-add-member", self.GROUP_2,
+             f"--groups={self.GROUP}"]
+        )
+        self.master.run_command(["ipa", "hbacrule-add", self.HBAC_RULE])
+        self.master.run_command(
+            ["ipa", "hbacrule-add-user", self.HBAC_RULE,
```

Comment:
This can be moved to tasks.py
add function hbacrule_add() like group_add()
add function: hbacrule_add_member() like group_add_member()